nomad/vault.go

```diff
@@ -0,0 +1,423 @@
+	}
+
+	// Get the Vault API configuration
+	apiConf, err := c.ApiConfig()
+	if err != nil {
+		return nil, fmt.Errorf("Failed to create Vault API config: %v", err)
+	}
+
+	// Create the Vault API client
+	client, err := vapi.NewClient(apiConf)
+	if err != nil {
+		v.logger.Printf("[ERR] vault: failed to create Vault client. Not retrying: %v", err)
+		return nil, err
+	}
+
+	// Set the token and store the client
+	client.SetToken(v.config.Token)
+	v.client = client
+	v.auth = client.Auth().Token()
+
+	// Prepare and launch the token renewal goroutine
+	v.shutdownCh = make(chan struct{})
+	go v.establishConnection()
+	return v, nil
+}
+
+// establishConnection is used to make first contact with Vault. This should be
+// called in a go-routine since the connection is retried til the Vault Client
+// is stopped or the connection is successfully made at which point the renew
+// loop is started.
+func (v *vaultClient) establishConnection() {
+	v.l.Lock()
+	v.establishingConn = true
+	v.l.Unlock()
+
+	// Create the retry timer and set initial duration to zero so it fires
+	// immediately
+	retryTimer := time.NewTimer(0)
+
+OUTER:
+	for {
+		select {
+		case <-v.shutdownCh:
+			return
+		case <-retryTimer.C:
+			// Ensure the API is reachable
+			if _, err := v.client.Sys().InitStatus(); err != nil {
+				v.logger.Printf("[WARN] vault: failed to contact Vault API. Retrying in %v",
+					v.config.ConnectionRetryIntv)
+				retryTimer.Reset(v.config.ConnectionRetryIntv)
+				continue OUTER
+			}
+
+			break OUTER
+		}
+	}
+
+	v.l.Lock()
+	v.connEstablished = true
+	v.establishingConn = false
+	v.l.Unlock()
+
+	// Retrieve our token, validate it and parse the lease duration
+	if err := v.parseSelfToken(); err != nil {
+		v.logger.Printf("[ERR] vault: failed to lookup self token and not retrying: %v", err)
+		return
+	}
+
+	// Set the wrapping function such that token creation is wrapped now
+	// that we know our role
+	v.client.SetWrappingLookupFunc(v.getWrappingFn())
+
+	// If we are given a non-root token, start renewing it
+	if v.token.Root {
+		v.logger.Printf("[DEBUG] vault: not renewing token as it is root")
+	} else {
+		v.logger.Printf("[DEBUG] vault: token lease duration is %v",
+			time.Duration(v.token.CreationTTL)*time.Second)
+		go v.renewalLoop()
+	}
+}
+
+// renewalLoop runs the renew loop. This should only be called if we are given a
+// non-root token.
+func (v *vaultClient) renewalLoop() {
+	v.l.Lock()
+	v.renewalRunning = true
+	v.l.Unlock()
+
+	// Create the renewal timer and set initial duration to zero so it fires
+	// immediately
+	authRenewTimer := time.NewTimer(0)
+
+	// Backoff is to reduce the rate we try to renew with Vault under error
+	// situations
+	backoff := 0.0
+
+	for {
+		select {
+		case <-v.shutdownCh:
+			return
+		case <-authRenewTimer.C:
+			// Renew the token and determine the new expiration
+			err := v.renew()
+			currentExpiration := v.lastRenewed.Add(time.Duration(v.token.CreationTTL) * time.Second)
```

Comment:
The TTL of the token will be half the CreationTTL by the time Nomad renews it. But the increment for the renewal will always be CreationTTL and the response should contain the same value (+/- a second or two considering network delays). That's it.